### ISSUE TITLE
Polars built in pyodide ?

### ISSUE THREAD

**Issue by oscar6echo** [pola-rs/polars#3672]
Could polars be compiled in [pyodide](https://pyodide.org/en/stable/index.html), like all [these packages](https://pyodide.org/en/stable/usage/packages-in-pyodide.html), and thus be easily available in [jupyterlite](https://jupyterlite.readthedocs.io/en/latest/) ?

The use case:
I am a heavy pandas user, will soon try and replace it with polars for new projects. I am also increasingly using jupyterlite to allow users to run "custom analytics" autonomously - with zero install. Naturally in a browser, datasets are unlikely to be really big (where polars shine), but if they are made to run both in the browser and locally (which I intend to do), it would be convenient.

From a distance, it would seem feasible as rust and wasm are not stranger to each other, correct ?

I am floating the idea to see what other people and the polars contributors think.
To polars creator/contributors: By the way you then demo polars live.

Last, but not least, congratulations for this impressive lib, and the neat docs ! :clap:

**Comment by ghuls** [comment]
Pyodide doesn't support threads AFAIK, which are currently needed by polars.

**Comment by oscar6echo** [comment]
Ok thank you for the clarification.  

Closing as answered.

**Comment by kylebarron** [comment]
I did some research on this tonight and it seems plausible to get a working pyodide build, even without a separate polars wasm build.

The first thing to note is that `pyodide` builds rust code against the **`wasm32-unknown-emscripten`** target _not_ `wasm32-unknown-unknown` target, which is where (I think) most of the existing wasm-related polars work has gone.

First, I was able to get a wasm build of [`py-geopolars`](https://github.com/kylebarron/geopolars/tree/master/py-geopolars) working in pyodide, though this [links only against the `Series` export](https://github.com/kylebarron/geopolars/blob/80613ece0360d45607602d47112ec322d834be4a/py-geopolars/src/ffi.rs#L2). This wasm loaded and was parsed successfully in pyodide, and then eventually failed when it tried to import and subclass from `polars.Series` on the Python side.

Then, turning to py-polars itself, I made some progress. I had to remove the `parquet`, `fmt`, `decompress-fast`, and `extract_jsonpath` features from polars, due to various dependency errors that I didn't want to spend time debugging at the moment.

After compiling for an hour, using the steps below, I hit only a couple errors:

1. a couple now-erroneous imports from features I removed, which should be easy to handle
2. Hundreds of `error: #[ctor] is not supported on the current target` errors, like this

    ```
    2310   │ error: #[ctor] is not supported on the current target
    2311   │   --> src/lazy/dataframe.rs:95:1
    2312   │    |
    2313   │ 95 | #[pymethods]
    2314   │    | ^^^^^^^^^^^^
    2315   │    |
    2316   │    = note: this error originates in the attribute macro `$crate::ctor` (in Nightly builds, run with -Z macro-backtrac
           │ e for more info)
    ```

    which I don't understand. Given this, it _seems_ like if the `#[ctor]` issue were solved, a polars pyodide build could be achieved by just adding a new feature flag to the `py-polars` crate.

	It seems like this might be an issue with python classes. I'll ask on the pyo3 issue tracker.

#### Repro for build attempt

My build steps (derived from [here](https://github.com/josephrocca/tokenizers-pyodide#build-instructions)). My branch is [here](https://github.com/kylebarron/polars/tree/kyle/pyodide).

Install nightly
```bash
rustup toolchain install nightly
rustup target add --toolchain nightly wasm32-unknown-emscripten
```

Install maturin `0.13` to take advantage of its [new support for emscripten](https://github.com/PyO3/maturin/pull/974). Note that maturin even [tests against emscripten on CI](https://github.com/PyO3/maturin/blob/fbed43f499ae281ac693f93db06b7dbcb79dfb74/.github/workflows/test.yml#L150) now.

```bash
pip install -U maturin
```

Set up emscripten 3.1.14. As of pyodide 0.21.0a3, pyodide is compiled against emscripten 3.1.14 and any extension module must also be compiled against the same version.

```
git clone https://github.com/emscripten-core/emsdk.git # tested and working at commit hash 961e66c
cd emsdk
./emsdk install 3.1.14
./emsdk activate 3.1.14
source ./emsdk_env.sh
```

Compile polars. Note this requires python 3.10; I used 3.10.5.

```bash
# cd into python dir
cd ../py-polars
RUSTUP_TOOLCHAIN=nightly maturin build --release -o dist --target wasm32-unknown-emscripten -i python3.10
```

#### Threads

> Pyodide doesn't support threads AFAIK, which are currently needed by polars.

In terms of threading, it looks like [a previous rust package was able to fix this](https://github.com/pyodide/pyodide/issues/2816#issuecomment-1172030538) just by turning changing the number of threads. Seems like it would be straightforward to do the same by setting `RAYON_NUM_THREADS=1`.

#### References/Further reading

- https://github.com/josephrocca/tokenizers-pyodide#build-instructions
- https://github.com/PyO3/pyo3/issues/2412
- https://github.com/pyodide/pyodide/issues/1973

**Comment by kylebarron** [comment]
From https://github.com/PyO3/pyo3/issues/2517#issuecomment-1188629768, looks like a pyodide build would be possible if the polars python binding switched to using a single `#[pymethods]` implementation.

I need to read up on rust macros; if the macros could be used _inside_ a single 
```rs
#[pymethods]
impl PySeries {
```
instead of being called at the top level, seems like it might be possible to switch to this?

**Comment by jonashaag** [comment]
This is still blocking a WASM build. I'll have a look at merging those blocks.

**Comment by jonashaag** [comment]
With merged blocks, got this https://github.com/pyodide/pyodide/issues/4016

**Comment by bitsondatadev** [comment]
Didn't see this linked anywhere and feel it's also relevant as new rollouts for thread support in WA/Pyosdide could unblock something.

https://github.com/pyodide/pyodide/issues/237

**Comment by lorentzenchr** [comment]
@stinodego @alexander-beedie @ghuls @MarcoGorelli Would you consider to reopen this issue or is it indeed out of scope?

**Comment by astrojuanlu** [comment]
xref https://github.com/pyodide/pyodide/issues/4611

**Comment by georgestagg** [comment]
I'm taking a shot at this. I have an LLVM fork that can work with multiple uses of `#[pymethods]`. More info at https://github.com/dtolnay/inventory/issues/71#issuecomment-2391782635 and https://github.com/llvm/llvm-project/pull/111008#issue-2564381414.

Using the LLVM fork with a custom `rustc`, disabling a bunch of features in Polars, and making some patches for wasm32 & Emscripten - (my patches are at https://github.com/georgestagg/polars/tree/emscripten) - I can build a wheel for Pyodide:

![371104753-f79226d6-c58b-47d7-af16-27f00b380488](https://github.com/user-attachments/assets/34656515-f1ed-434d-94cf-ba982a60e221)

The path to making this work properly is still a long one, but looks as follows:

1) Upstream LLVM patch to fix multiple `.init_array` symbols in Wasm objects.
2) Wait for a version of `rustc` to be released with the LLVM fixes included.
3) Update the inventory package to add `.init_array` symbols under WebAssembly.
4) Upstream patches to Polars to support the `wasm32-unknown-emscripten` target.

**Comment by ritchie46** [comment]
https://github.com/pola-rs/polars/pull/19937#issuecomment-2515706991

**Comment by lorentzenchr** [comment]
Solved in #20383.
